plugins/formal-verification/formver.plugin/src/org/jetbrains/kotlin/formver/reporting/ErrorReporter.kt

```diff
@@ -0,0 +1,121 @@
+import org.jetbrains.kotlin.fir.types.ConeKotlinType
+import org.jetbrains.kotlin.fir.types.renderReadable
+import org.jetbrains.kotlin.formver.PluginErrors
+import org.jetbrains.kotlin.formver.embeddings.SourceRole
+import org.jetbrains.kotlin.formver.embeddings.SourceRole.ParamFunctionLeakageCheck.fetchLeakingFunction
+import org.jetbrains.kotlin.formver.viper.errors.VerificationError
+
+/**
+ * Abstract class defining a [report] method used to warn users with human-readable messages.
+ */
+abstract class ErrorReporter<S : SourceRole>(
+    protected val error: VerificationError,
+    protected val role: S,
+) {
+    abstract fun report(reporter: DiagnosticReporter, source: KtSourceElement?, context: CheckerContext)
+}
+
+class ReturnsEffectReporter(error: VerificationError, source: SourceRole.ReturnsEffect) :
+    ErrorReporter<SourceRole.ReturnsEffect>(error, source) {
+    private val SourceRole.ReturnsEffect.asUserFriendlyMessage: String
+        get() = when (this) {
+            is SourceRole.ReturnsEffect.Bool -> if (bool) "false" else "true"
+            is SourceRole.ReturnsEffect.Null -> if (negated) "null" else "non-null"
+            else -> throw IllegalStateException("Unknown returns effect: $this")
+        }
+
+    override fun report(reporter: DiagnosticReporter, source: KtSourceElement?, context: CheckerContext) {
+        reporter.reportOn(source, PluginErrors.UNEXPECTED_RETURNED_VALUE, role.asUserFriendlyMessage, context)
+    }
+}
+
+class CallsInPlaceReporter(error: VerificationError, role: SourceRole.CallsInPlaceEffect) :
+    ErrorReporter<SourceRole.CallsInPlaceEffect>(error, role) {
+    private val SourceRole.CallsInPlaceEffect.asUserFriendlyMessage: String
+        get() = when (kind) {
+            EventOccurrencesRange.AT_MOST_ONCE -> "at most once"
+            EventOccurrencesRange.EXACTLY_ONCE -> "exactly once"
+            EventOccurrencesRange.AT_LEAST_ONCE -> "at least once"
+            EventOccurrencesRange.MORE_THAN_ONCE -> "more than once"
+            else -> throw IllegalStateException("Unknown kind of range: $kind")
+        }
+
+    override fun report(reporter: DiagnosticReporter, source: KtSourceElement?, context: CheckerContext) {
+        reporter.reportOn(source, PluginErrors.INVALID_INVOCATION_TYPE, role.paramSymbol, role.asUserFriendlyMessage, context)
+    }
+}
+
+class LeakingLambdaReporter(error: VerificationError, role: SourceRole.ParamFunctionLeakageCheck) :
+    ErrorReporter<SourceRole.ParamFunctionLeakageCheck>(error, role) {
+    override fun report(reporter: DiagnosticReporter, source: KtSourceElement?, context: CheckerContext) {
+        reporter.reportOn(source, PluginErrors.LAMBDA_MAY_LEAK, error.fetchLeakingFunction(), context)
+    }
+}
+
+class ConditionalEffectReporter(error: VerificationError, role: SourceRole.ConditionalEffect) :
+    ErrorReporter<SourceRole.ConditionalEffect>(error, role) {
+    private inline val ConeKotlinType.rendered: String
```

Comment:
A lot of these functions seem like general utilities that would be better off not tied to a class.